File: go/bcm.go
package bcm

import (
	"bytes"
	"encoding/binary"
	"encoding/hex"
	"fmt"

	"github.com/libsv/go-bt/v2"
)

const TxChunkPrefix = "TXBYTES."
const KUMChunkPrefix = "TXOKEYS."

type KeyUsageMetadata struct {
	Vout           uint32 `json:"vout"`
	ScriptType     string `json:"script_type"`
	KeyFingerprint string `json:"key_fingerprint"`
	KeyDerivation  string `json:"key_derivation"`
}

type BitCacheMessage struct {
	Tx   *bt.Tx              `json:"tx"`
	KUMs []*KeyUsageMetadata `json:"key_map"`
}

func NewBitCacheMessage(bcmBytes []byte) (*BitCacheMessage, error) {
	txChunkPrefix := bcmBytes[:8]
	if !bytes.Equal(txChunkPrefix, []byte(TxChunkPrefix)) {
		return nil, fmt.Errorf("invalid tx chunk prefix")
	}

	txChunkSize := binary.LittleEndian.Uint64(bcmBytes[8:16])

	txBytes := bcmBytes[16 : 16+txChunkSize]
	tx, err := bt.NewTxFromBytes(txBytes)
	if err != nil {
		return nil, err
	}

	cursor := 16 + txChunkSize
	kumChunkPrefix := bcmBytes[cursor : cursor+8]
	if !bytes.Equal(kumChunkPrefix, []byte(KUMChunkPrefix)) {
		return nil, fmt.Errorf("invalid KUM chunk prefix")
	}

	cursor += 8
	kumChunkSize := binary.LittleEndian.Uint64(bcmBytes[cursor : cursor+8])

	cursor += 8
	kumBytes := bcmBytes[cursor : cursor+kumChunkSize]

	kums := []*KeyUsageMetadata{}
	kcursor := uint64(0)
	for kcursor < kumChunkSize {
		vout := binary.LittleEndian.Uint32(kumBytes[kcursor : kcursor+4])
		kcursor += 4

		scriptType, bytesRead := readLengthPrefixed(kumBytes, kcursor)
		kcursor += bytesRead
		kfp, bytesRead := readLengthPrefixed(kumBytes, kcursor)
		kcursor += bytesRead
		kd, bytesRead := readLengthPrefixed(kumBytes, kcursor)
		kcursor += bytesRead

		kum := &KeyUsageMetadata{
			Vout:           vout,
			ScriptType:     string(scriptType),
			KeyFingerprint: hex.EncodeToString(kfp),
			KeyDerivation:  string(kd),
		}
		kums = append(kums, kum)
	}

	return &BitCacheMessage{
		Tx:   tx,
		KUMs: kums,
	}, nil
}

func readLengthPrefixed(data []byte, cursor uint64) ([]byte, uint64) {
	lengthPrefixSize := uint64(2)
	length := binary.LittleEndian.Uint16(data[cursor : cursor+lengthPrefixSize])
	strStart := cursor + lengthPrefixSize
	strEnd := strStart + uint64(length)
	return data[strStart:strEnd], lengthPrefixSize + uint64(length)
}

func (bcm *BitCacheMessage) Bytes() ([]byte, error) {
	var buffer bytes.Buffer

	// Write TxChunkPrefix
	buffer.Write([]byte(TxChunkPrefix))

	// Serialize the transaction and write its length and bytes
	txBytes := bcm.Tx.Bytes()
	txChunkSize := uint64(len(txBytes))
	txSizeBytes := make([]byte, 8)
	binary.LittleEndian.PutUint64(txSizeBytes, txChunkSize)
	buffer.Write(txSizeBytes)
	buffer.Write(txBytes)

	// Write KUMChunkPrefix
	buffer.Write([]byte(KUMChunkPrefix))

	// Calculate KUM chunk size and write it later
	kumSizePos := buffer.Len()
	buffer.Write(make([]byte, 8)) // Placeholder for KUM chunk size

	// Serialize each KeyUsageMetadata and write to buffer
	for _, kum := range bcm.KUMs {
		voutBytes := make([]byte, 4)
		binary.LittleEndian.PutUint32(voutBytes, kum.Vout)
		buffer.Write(voutBytes)

		writeLengthPrefixed(&buffer, []byte(kum.ScriptType))
		kfpBytes, err := hex.DecodeString(kum.KeyFingerprint)
		if err != nil {
			return nil, err
		}
		writeLengthPrefixed(&buffer, kfpBytes)
		writeLengthPrefixed(&buffer, []byte(kum.KeyDerivation))
	}

	// Update KUM chunk size
	kumChunkSize := uint64(buffer.Len() - kumSizePos - 8)
	kumSizeBytes := make([]byte, 8)
	binary.LittleEndian.PutUint64(kumSizeBytes, kumChunkSize)
	copy(buffer.Bytes()[kumSizePos:], kumSizeBytes)

	return buffer.Bytes(), nil
}

func writeLengthPrefixed(buffer *bytes.Buffer, data []byte) {
	lengthBytes := make([]byte, 2)
	binary.LittleEndian.PutUint16(lengthBytes, uint16(len(data)))
	buffer.Write(lengthBytes)
	buffer.Write(data)
}

File: ts/bcm.ts
import { Buffer } from 'buffer';

const TxChunkPrefix = "TXBYTES.";
const KUMChunkPrefix = "TXOKEYS.";

interface KeyUsageMetadata {
    vout: number;
    script_type: string;
    key_fingerprint: string;
    key_derivation: string;
}

class BitCacheMessage {
    Tx: string;
    KUMs: KeyUsageMetadata[];

    constructor(tx: string, kums: KeyUsageMetadata[]) {
        this.Tx = tx;
        this.KUMs = kums;
    }

    static fromBytes(bcmBytes: Buffer): BitCacheMessage {
        if (bcmBytes.slice(0, 8).toString() !== TxChunkPrefix) {
            throw new Error("Invalid tx chunk prefix");
        }

        const txChunkSize = bcmBytes.readUInt32LE(8);
        const txBytes = bcmBytes.slice(16, 16 + txChunkSize);

        let cursor = 16 + txChunkSize;
        if (bcmBytes.slice(cursor, cursor + 8).toString() !== KUMChunkPrefix) {
            throw new Error("Invalid KUM chunk prefix");
        }

        cursor += 8;
        const kumChunkSize = bcmBytes.readUInt32LE(cursor);
        cursor += 8;

        const kumBytes = bcmBytes.slice(cursor, cursor + kumChunkSize);

        let kcursor = 0;
        const kums: KeyUsageMetadata[] = [];
        while (kcursor < kumChunkSize) {
            const vout = kumBytes.readUInt32LE(kcursor);
            kcursor += 4;

            const [scriptType, bytesReadST] = readLengthPrefixed(kumBytes, kcursor);
            kcursor += bytesReadST;

            const [kfp, bytesReadKFP] = readLengthPrefixed(kumBytes, kcursor);
            kcursor += bytesReadKFP;

            const [kd, bytesReadKD] = readLengthPrefixed(kumBytes, kcursor);
            kcursor += bytesReadKD;

            kums.push({
                vout: vout,
                script_type: scriptType.toString(),
                key_fingerprint: kfp.toString('hex'),
                key_derivation: kd.toString(),
            });
        }

        return new BitCacheMessage(txBytes.toString('hex'), kums);
    }

    toBytes(): Buffer {
        let buffer = Buffer.alloc(1000); // Start with a buffer of 1000 bytes
        let offset = 0;

        offset += buffer.write(TxChunkPrefix, offset);
        const txLengthBigInt = BigInt(this.Tx.length / 2); // Convert hex string length to byte length
        buffer = ensureBufferSize(buffer, offset, 8);
        offset = writeUInt64LE(buffer, txLengthBigInt, offset);

        const tx = Buffer.from(this.Tx, 'hex');
        buffer = ensureBufferSize(buffer, offset, tx.length);
        offset += tx.copy(buffer, offset);

        offset += buffer.write(KUMChunkPrefix, offset);
        const kumSizePos = offset;
        offset += 8; // Reserve space for KUM chunk size

        this.KUMs.forEach(kum => {
            buffer = ensureBufferSize(buffer, offset, 4);
            buffer.writeUInt32LE(kum.vout, offset);
            offset += 4;

            const scriptTypeBuffer = Buffer.from(kum.script_type);
            buffer = ensureBufferSize(buffer, offset, scriptTypeBuffer.length + 2);
            offset = writeLengthPrefixed(buffer, scriptTypeBuffer, offset);

            const kfpBuffer = Buffer.from(kum.key_fingerprint, 'hex');
            buffer = ensureBufferSize(buffer, offset, kfpBuffer.length + 2);
            offset = writeLengthPrefixed(buffer, kfpBuffer, offset);

            const kdBuffer = Buffer.from(kum.key_derivation);
            buffer = ensureBufferSize(buffer, offset, kdBuffer.length + 2);
            offset = writeLengthPrefixed(buffer, kdBuffer, offset);
        });

        const kumChunkSize = offset - kumSizePos - 8;
        buffer = ensureBufferSize(buffer, kumSizePos, 4);
        buffer.writeUInt32LE(kumChunkSize, kumSizePos); // Write the actual KUM chunk size

        return buffer.slice(0, offset);
    }
}

function readLengthPrefixed(data: Buffer, cursor: number): [Buffer, number] {
    const length = data.readUInt16LE(cursor);
    const strStart = cursor + 2;
    const strEnd = strStart + length;
    return [data.slice(strStart, strEnd), length + 2];
}

function writeLengthPrefixed(buffer: Buffer, data: Buffer, offset: number): number {
    buffer.writeUInt16LE(data.length, offset);
    offset += 2;
    data.copy(buffer, offset);
    return offset + data.length;
}

function writeUInt64LE(buffer: Buffer, value: bigint, offset: number): number {
    const MAX_UINT32 = BigInt(0xFFFFFFFF);
    const bigValue = value / MAX_UINT32;
    const lowValue = value % MAX_UINT32;

    buffer.writeUInt32LE(Number(lowValue), offset); // Write the low 32 bits
    buffer.writeUInt32LE(Number(bigValue), offset + 4); // Write the high 32 bits

    return offset + 8; // Return the new offset
}

function ensureBufferSize(buffer: Buffer, currentOffset: number, additionalLength: number): Buffer {
    if (currentOffset + additionalLength > buffer.length) {
        const newBuffer = Buffer.alloc(buffer.length + additionalLength); // Increase the size
        buffer.copy(newBuffer);
        return newBuffer;
    }
    return buffer;
}

export { BitCacheMessage, KeyUsageMetadata };
